src/celeritas/neutron/interactor/detail/CascadeCollider.hh

```diff
@@ -0,0 +1,235 @@
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the final state of outgoing particles from the two-body intra-nucleus
+ * nucleon-nucleon collision in the center of mass (c.m.) frame and return
+ * them after converting momentum to the lab frame. This performs the same
+ * sampling routine as in Geant4's G4ElementaryParticleCollider, mainly
+ * implemented in collide and generateSCMfinalState methods. The cos\theta
+ * distribution in c.m. is inversely sampled using the tabulated cumulative
+ * distribution function (c.d.f) data in the kinetic energy and cos\theta bins
+ * which are implemented in G4CascadeFinalStateAlgorithm::GenerateTwoBody and
+ * G4NumIntTwoBodyAngDst::GetCosTheta methods.
+ */
+template<class Engine>
+CELER_FUNCTION auto CascadeCollider::operator()(Engine& rng) -> FinalState
+{
+    // Sample cos\theta of outgoing particles in the center of mass frame
+    real_type cdf = generate_canonical(rng);
+    TwodGridData cdf_grid = shared_.angular_cdf[ch_id_];
+    Grid energy_grid(cdf_grid.x, shared_.reals);
+    real_type cos_theta{0};
+
+    if (kin_energy_ < energy_grid.back())
+    {
+        // Find cos\theta from tabulated angular data for a given c.d.f.
+        Grid cos_grid(cdf_grid.y, shared_.reals);
+        TwodGridCalculator calc_cdf(cdf_grid, shared_.reals);
+
+        size_type idx = cos_grid.size() - 2;
+        real_type cdf_upper = 0;
+        real_type cdf_lower = 1;
+
+        do
+        {
+            cdf_upper = cdf_lower;
+            cdf_lower = calc_cdf({kin_energy_, cos_grid[idx]});
+        } while (cdf_lower > cdf && idx-- > 0);
+
+        real_type frac = (cdf - cdf_lower) / (cdf_upper - cdf_lower);
+        cos_theta = fma(frac, cos_grid[idx + 1] - cos_grid[idx], cos_grid[idx]);
+    }
+    else
+    {
+        // Sample the angle outside tabulated data (unlikely)
+        real_type slope = 2 * energy_grid.back() * ipow<2>(cm_p_);
+        cos_theta = std::log(1 + cdf * std::expm1(2 * slope)) / slope - 1;
+    }
+
+    // Sample the momentum of outgoing particles in the center of mass frame
+    Real3 mom = cm_p_ * from_spherical(cos_theta, sample_phi_(rng));
+
+    // Rotate the momentum along the reference z-axis
+    FourVector fv = {mom,
+                     std::sqrt(dot_product(mom, mom)
+                               + ipow<2>(value_as<MevMass>(bullet_.mass)))};
+
+    // Find the final state of outgoing particles
+    FinalState result = {bullet_, target_};
+
+    FourVector cm_momentum = target_.four_vec;
+    boost(-cm_velocity_, &cm_momentum);
+
+    Real3 cm_dir = make_unit_vector(-cm_momentum.mom);
+    real_type vel_parallel = dot_product(cm_velocity_, cm_dir);
+
+    // Degenerated if velocity perpendicular to the c.m. momentum is small
+    bool degenerated
+        = (dot_product(cm_velocity_, cm_velocity_) - ipow<2>(vel_parallel)
+           < this->epsilon());
+
+    if (!degenerated)
```

Comment:
Are you sure this expression is right? It's comparing the difference in the square of speeds? Maybe it would be better to calculate the `vscm` and then check the magnitude of that...

Also it appears that `epsilon` here needs to have units of `[1/c^2]`? Is that right?

Reply:
I agree that they should be equivalent, i.e., `norm(vscm) == sqrt(dot_product(cm_velocity_, cm_velocity_) - ipow<2>(vel_parallel))`.  epsilon is small arbitrary magic number from Geant4 and used for condition for the degeneracy [1/c^2] and vscm [1/c] as c = 1 in Geant4.  Probably we keep the expressions as they are for now and refactor them later.  Anyway, I will add 1/c^2 in epsilon.